sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -77,4 +77,29 @@ package object expressions  {
     /** Uses the given row to store the output of the projection. */
     def target(row: MutableRow): MutableProjection
   }
+
+  /**
+   * A utility classes for check the expression equality in semantic, not literally.
+   */
+  trait ExpressionEquals
+
+  private[this] class AttributeRefEquals(private val normalized: Expression)
+    extends ExpressionEquals {
+    override def equals(other: Any): Boolean = other match {
+      case o: AttributeRefEquals => normalized == o.normalized
+      case _ => false
+    }
+
+    override def hashCode(): Int = normalized.hashCode()
+  }
+
+  object ExpressionEquals {
```

Comment:
Why not follow the pattern already created by `AttributeSet` and create an `ExpressionSet`/`ExpressionMap` (with tests)?